src/validate_async.cc

```diff
@@ -6,69 +6,130 @@ using std::string;
 using namespace v8;
 using namespace protagonist;
 
-// Async Validate
-void AsyncValidate(uv_work_t* request);
+using Nan::AsyncQueueWorker;
+using Nan::AsyncWorker;
+using Nan::Callback;
+using Nan::HandleScope;
 
-// Async Validate Handler
-void AsyncValidateAfter(uv_work_t* request);
+namespace {
 
-// Threadpooling libuv baton
-struct ValidateBaton {
-
-    ~ValidateBaton(){
-        if (sourceData) {
-            delete sourceData;
-        }
-        if (result) {
-            drafter_free_result(result);
-        }
-    }
-    // Callback
-    Nan::Persistent<Function> callback;
-
-    // Input
+class ValidateWorker : public AsyncWorker {
     drafter_parse_options parseOptions;
+    drafter_serialize_options serializeOptions;
     Nan::Utf8String *sourceData;
 
     // Output
     drafter_result *result;
     int parse_err_code;
+
+   public:
+    ValidateWorker(Callback *callback, drafter_parse_options parseOptions,
+                   Nan::Utf8String *sourceData)
+        : AsyncWorker(callback),
+          parseOptions(parseOptions),
+          sourceData(sourceData),
+          result(nullptr),
+          parse_err_code(-1) {}
+
+    void Execute() {
+        parse_err_code =
+            drafter_check_blueprint(*(*sourceData), &result, parseOptions);
+    }
+
+    void HandleOKCallback() {
+        Nan::HandleScope scope;
+
+        // Error Object
+        Local<Value> argv[2];
+
+        switch (parse_err_code) {
+            case DRAFTER_EUNKNOWN:
+                argv[0] = Nan::Error("Parser: Unknown Error");
+                argv[1] = Nan::Null();
+                break;
+            case DRAFTER_EINVALID_INPUT:
+                argv[0] = Nan::Error("Parser: Invalid Input");
+                argv[1] = Nan::Null();
+                break;
+            case DRAFTER_EINVALID_OUTPUT:
+                argv[0] = Nan::Error("Parser: Invalid Output");
+                argv[1] = Nan::Null();
+                break;
+            default:
+                argv[0] = Nan::Null();
+                if (result) {
+                    argv[1] =
+                        refract2v8(result, {true, DRAFTER_SERIALIZE_JSON});
+                } else {
+                    argv[1] = Nan::Null();
+                }
+        }
+
+        // v8::Local<v8::Value> argv[] = {
+        //    Nan::Null(),
+        //    refract2v8(result, serializeOptions)
+        //};
+
+        // if (0 != parse_err_code) {
+        //    argv[0] = annotations2v8(result);
+        //}
```

Comment:
Let's remove this.

Reply:
Sry, I remove it, but it seem I forgot to push it